### PR TITLE
feature/344: Fetch nodes and processes cpu and mem usage using metrics api

### PR DESCRIPTION
**Pull request by nightfury1204**
This PR provides backend support for : https://github.com/convox/issues-private/issues/344

It fetches the nodes and processes memory and cpu usage using metrics api. mertic-server installed in k8s cluster is responsible for that.

### CODE REVIEW

File: provider/k8s/helpers.go
Comment (Twsouza) [inline provider/k8s/helpers.go:329]: This function is not meant to be a helper, it's a provider function, perhaps it's better to move it to be a Provider function, something like `p.CalculatePodCpuAndMem`. Since you have the metrics interface in the Provider struct, you won't need to pass it to the function.

**Review by Twsouza — CHANGES_REQUESTED** [review]
It's only returning the sum of the pods' metrics. It would be good to have another way to get the nodes' metrics (even in a different endpoint).

File: go.mod
Comment (nightfury1204) [reply go.mod:58]: this version is tight to the client-go version which is `v0.16.6`, to update it to latest version we need to update all the following related library too:
```
        k8s.io/api v0.16.6
	k8s.io/apimachinery v0.16.6
	k8s.io/client-go v0.16.6
	k8s.io/code-generator v0.16.6
```
Also codes needs to be updated too to use these updated packages, not sure how much though. That's why i didn't use the latest version.
I think it should be done in a seperate PR, because it could be a major change.

**Comment by nightfury1204** [comment]
> It's only returning the sum of the pods' metrics. It would be good to have another way to get the nodes' metrics (even in a different endpoint).

@Twsouza it actually doesn't return the sum of the pods' metrics, it's returning sum of container metrics for each pod, because a pod can have multiple containers. This pr doesn't provide implementation for nodes metrics.

File: provider/k8s/helpers.go
Comment (nightfury1204) [reply provider/k8s/helpers.go:329]: `m *metricsv1beta1.PodMetrics` this parameter is struct, it contains containers usage metrics. it differs pod to pod, so i have to pass it as parameter. It basically sums the resource usage and return it. That's why i put it in the helpers method.
But it can be made provider function too, but the parameter still will be there.
Should i make it provider function, what's your call?

Comment (Twsouza) [reply provider/k8s/helpers.go:329]: No need, leave as it is 👍

**Review by Twsouza — APPROVED** [review]
👍

File: provider/k8s/process.go
Comment (heronrs) [inline provider/k8s/process.go:164]: I think we can return the error here. It could be useful to notify the client about problems. We include metrics server by default in the rack instalation so if it's missing we'd know.

File: provider/k8s/process.go
Comment (heronrs) [inline provider/k8s/process.go:171]: just checking. process.Id here is equal the pod name, right?

**Comment by Twsouza** [comment]
@nightfury1204 is this only applied to processes? how is the instance going to be applied?

**Comment by heronrs** [comment]
> @nightfury1204 is this only applied to processes? how is the instance going to be applied?

@nightfury1204 I think we'll need a similar endpoint for getting the metrics of the instance nodes as it's one of the tabs in the console UI.

It's fine if you want to do it in a different PR.

File: provider/k8s/process.go
Comment (nightfury1204) [reply provider/k8s/process.go:171]: yes, in the existing code I saw using pod name used as id. ref: https://github.com/convox/convox/blob/master/provider/k8s/process.go#L591

**Comment by nightfury1204** [comment]
> > @nightfury1204 is this only applied to processes? how is the instance going to be applied?
> 
> @nightfury1204 I think we'll need a similar endpoint for getting the metrics of the instance nodes as it's one of the tabs in the console UI.
> 
> It's fine if you want to do it in a different PR.

Got it. In that case, I will add the node usages metrics in this PR. That way we all be in the same context and they will get merged together.
I am putting label **do not merge** label for now. Once I pushed the changes I will remove it.

cc: @heronrs @Twsouza

File: provider/k8s/helpers.go
Comment (nightfury1204) [reply provider/k8s/helpers.go]: addressed it in next commit

File: provider/k8s/process.go
Comment (nightfury1204) [reply provider/k8s/process.go]: addressed it in the next commit

File: pkg/cli/instances.go
Comment (heronrs) [inline pkg/cli/instances.go:46]: #nit you can remove this comment as this repo only contains the code for v3 racks.

**Review by heronrs — APPROVED** [review]
LGTM. minor nit pick

File: pkg/cli/instances.go
Comment (nightfury1204) [reply pkg/cli/instances.go]: removed it

**Review by heronrs — APPROVED** [review]
lgtm :fire:

**Review by Twsouza — APPROVED** [review]
🚢

**Pull request merged**
